src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -175,11 +175,256 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Bindin
+            bool found = TryLookupLengthOrCount(inputType, lookupResult, out lengthProperty, ref useSiteInfo);
+            lookupResult.Free();
+            return found;
+        }
+
+        private BoundListPatternClause BindListPatternClause(
+            PropertyPatternClauseSyntax node,
+            TypeSymbol inputType,
+            bool permitDesignations,
+            ref PropertySymbol? lengthProperty,
+            ref bool hasErrors,
+            BindingDiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.IsKind(SyntaxKind.ListPatternClause));
+
+            TypeSymbol elementType;
+            BoundIndexerAccess? indexerAccess = null;
+            PropertySymbol? indexerSymbol = null;
+            bool isUnsupported = false;
+            if (inputType.IsSZArray())
+            {
+                elementType = ((ArrayTypeSymbol)inputType).ElementType;
+            }
+            else if (TryFindIndexerOrIndexerPattern(node, inputType, argIsIndex: true, out indexerAccess, out Symbol? patternSymbol, diagnostics))
+            {
+                if (patternSymbol is PropertySymbol indexer)
+                {
+                    indexerSymbol = indexer;
+                    elementType = indexer.Type;
+                }
+                else
+                {
+                    Debug.Assert(indexerAccess is not null);
+                    elementType = indexerAccess.Type;
+                }
+            }
+            else
+            {
+                elementType = CreateErrorType();
+                isUnsupported = true;
+            }
+
+            // We may have looked up the length property for the length pattern.
+            if (lengthProperty is null && !TryLookupLengthOrCount(node, inputType, out lengthProperty))
+                isUnsupported = true;
+
+            if (isUnsupported && !inputType.IsErrorType())
+            {
+                // Report an unsupported type if either the length property or a viable indexer is missing.
+                Error(diagnostics, ErrorCode.ERR_UnsupportedTypeForListPattern, node, inputType);
+                hasErrors = true;
+            }
+
+            ImmutableArray<BoundPattern> subpatterns = BindListPatternSubpatterns(node.Subpatterns, inputType, elementType, permitDesignations, ref hasErrors, out bool sawSlice, diagnostics);
+            return new BoundListPatternClause(node, subpatterns, sawSlice, indexerAccess, indexerSymbol, hasErrors);
+        }
+
+        private bool TryFindIndexerOrIndexerPattern(
+            SyntaxNode syntax, TypeSymbol receiverType, bool argIsIndex,
+            out BoundIndexerAccess? indexerAccess,
+            out Symbol? patternSymbol,
+            BindingDiagnosticBag diagnostics)
+        {
+            patternSymbol = null;
+            var lookupResult = LookupResult.GetInstance();
+            CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
```

Comment:
I think there's still a problem with use-site errors. See test below. Can be PROTOTYPE'd.

```
        [Fact]
        public void ListPattern_UseSiteErrorOnIndexerAndSlice()
        {
            var missing_cs = @"
public class Missing
{
}
";
            var missingRef = CreateCompilation(missing_cs, assemblyName: "missing")
                .EmitToImageReference();

            var lib2_cs = @"
public class C
{
    public int Length => 0;
    public Missing this[int i] => throw null;
    public Missing Slice(int i, int j) => throw null;
}
";
            var lib2Ref = CreateCompilation(lib2_cs, references: new[] { missingRef })
                .EmitToImageReference();

            var source = @"
class D
{
    void M(C c)
    {
        _ = c is { var item };
        _ = c is { ..var rest };
        var index = c[^1];
        var range = c[1..^1];
    }
}
";
            var compilation = CreateCompilation(new[] { source, TestSources.Index, TestSources.Range },
                references: new[] { lib2Ref }, parseOptions: TestOptions.RegularWithListPatterns);
            // PROTOTYPE missing diagnostics
            compilation.VerifyEmitDiagnostics(
                // (8,21): error CS0012: The type 'Missing' is defined in an assembly that is not referenced. You must add a reference to assembly 'missing, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'.
                //         var index = c[^1];
                Diagnostic(ErrorCode.ERR_NoTypeDef, "c[^1]").WithArguments("Missing", "missing, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null").WithLocation(8, 21),
                // (9,21): error CS0012: The type 'Missing' is defined in an assembly that is not referenced. You must add a reference to assembly 'missing, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'.
                //         var range = c[1..^1];
                Diagnostic(ErrorCode.ERR_NoTypeDef, "c[1..^1]").WithArguments("Missing", "missing, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null").WithLocation(9, 21)
                );


            var tree = compilation.SyntaxTrees.First();
            var model = compilation.GetSemanticModel(tree, ignoreAccessibility: false);
            var declarations = tree.GetRoot().DescendantNodes().OfType<VarPatternSyntax>().ToArray();

            verify(declarations[0], "item", "Missing?");
            verify(declarations[1], "rest", "Missing?");

            void verify(VarPatternSyntax declaration, string name, string expectedType)
            {
                var local = (ILocalSymbol)model.GetDeclaredSymbol(declaration.Designation)!;
                Assert.Equal(name, local.Name);
                Assert.Equal(expectedType, local.Type.ToTestDisplayString(includeNonNullable: true));
            }
        }
```

Reply:
Looks like the `BindIndexerOrIndexedPropertyAccess` I'm using also fallback to TryBindIndexOrRange, but `is BoundAccess` fails and we throw away any diagnostics. 

For indexer access it's working almost accidentally - we report the error while `TryBindIndexOrRange` succeeds. As a result, we've bound the fallback (just like list patterns) but reporting errors from the previous lookup (which is thrown away for list patterns)

Maybe we should depend on `BindIndexerOrIndexedPropertyAccess` for the fallback lookup in which case we'll need to create a BoundIndexer off of BoundIndexOrRangePatternIndexerAccess, maybe.

I'll leave this as a prototype comment so you can decide what is the correct fix here. Thanks.

Reply:
Will be addressed by https://github.com/dotnet/roslyn/pull/53784 (https://github.com/alrz/roslyn/commit/0fa1661a24df52f182527fb80cb81ec0cdbc87b9)